Addition of TanhExp activation function - issue #45929

sangireddysiva opened the pull request:
Addition of TanhExp activation function based on issue #45929 that leads changes to tensorflow/python/keras/activations.py and tensorflow/python/ops/nn_impl.py

sangireddysiva committed:
Addition of TanhExp activation function - issue #45929

sangireddysiva commented:
We are not sure what exactly oneDNN unit tests failed checks are talking about. I could see failure in compiling programs and some functionality tests. are these failed checks meant for our code changes?

ematejska commented:
The API owners are moving to a rotation based system instead of the meetings. Looks like @fchollet is already a reviewer and adding @manivaradarajan. Please review on behalf of the API Owners.

manivaradarajan commented on tensorflow/python/keras/activations.py:
```diff
@@ -369,6 +369,29 @@ def tanh(x):
+@keras_export('keras.activations.tanhexp')
+@dispatch.add_dispatch_support
+def tanhexp(x):
+  """TanhExponential activation function, `tanhexp(x) = x * tanh(exp(x))`.
```
Rephrase to "TanhExp activation function" for consistency. Please fix throughout.

sangireddysiva replied:
updated activations.py with suggested changes

manivaradarajan reviewed (changes requested):
Overall ok. Addressing API documentation only. Please wait for fchollet's review.

sangireddysiva committed:
Rephrased "TanhExponential" to "TanhExp"

Rephrased "TanhExponential activation function" to "TanhExp activation function" for consistency in documentation

sangireddysiva reviewed (commented):
Rephrased "TanhExponential" to "TanhExp" for consistency

gbaned commented:
@sangireddysiva  Can you please address Ubuntu Sanity errors? Thanks!

sangireddysiva committed:
Ubuntu Sanity error fixes - pylint max characters

Updated documentation to cope up with pylint max 80 characters based on errors during Ubuntu Sanity checks

sangireddysiva committed:
Ubuntu Sanity error fixes - pylint max characters

Updated documentation to cope up with pylint max 80 characters based on errors during Ubuntu Sanity checks

sangireddysiva commented:
> @sangireddysiva Can you please address Ubuntu Sanity errors? Thanks!

@gbaned , Adjusted the documentation to address the Ubuntu Sanity errors

sangireddysiva committed:
Ubuntu Sanity Pylink error fixes

sangireddysiva commented:
@gbaned  , Added Pylint disable line too long statement to address the Ubuntu Sanity errors

sangireddysiva committed:
Ubuntu CPU - Doctest error fixes

mihaimaruseac commented on tensorflow/python/keras/activations.py:
```diff
@@ -369,6 +369,34 @@ def tanh(x):
+@keras_export('keras.activations.tanhexp')
+@dispatch.add_dispatch_support
+def tanhexp(x):
+  # pylint: disable=line-too-long
```
Would be better to reformat the comment to not have lines that are too long.

sangireddysiva commented:
@gbaned , fixed Ubuntu CPU Doctest errors, thanks

mihaimaruseac commented on tensorflow/python/keras/activations.py:
```diff
@@ -369,6 +369,34 @@ def tanh(x):
+  TanhExp activation function which returns `x * tanh(exp(x))`.
+  It is a smooth, non-monotonic function with high convergence speed and with good accuracy
+  that consistently matches or outperforms ReLU on lightweight (or) mobile neural networks.
+  Example Usage:
```
Space above and below this line

mihaimaruseac commented on tensorflow/python/ops/nn_impl.py:
```diff
@@ -549,6 +549,44 @@ def grad(dy):
+@dispatch.add_dispatch_support
+@custom_gradient.custom_gradient
+def tanhexp(features):
+  # pylint: disable=g-doc-args
```
Reformat comment instead of this

mihaimaruseac commented on tensorflow/python/ops/nn_impl.py:
```diff
@@ -549,6 +549,44 @@ def grad(dy):
+  The TanhExp activation function was introduced in "TanhExp: A Smooth Activation Function
+  with High Convergence Speed for Lightweight Neural Networks"
+  [Xinyu et al. 2020](https://arxiv.org/pdf/2003.09855.pdf) 
+  Args:
```
Space above this line

mihaimaruseac commented on tensorflow/python/ops/nn_impl.py:
```diff
@@ -549,6 +549,44 @@ def grad(dy):
+  Args:
+    features: A `Tensor` representing preactivation values.
+    name: A name for the operation (optional).
+  Returns:
```
Space above this line

sangireddysiva committed:
Reformatted the Documentation for pylint and doctest fixes

sangireddysiva commented:
@mihaimaruseac addresses the requested changes to the documentation, thanks

sangireddysiva commented:
@gbaned , There are still some failed checks, do we need to look into them before merging? Thanks,

gbaned commented:
@mihaimaruseac Can you please assist on above comments from @sangireddysiva. Thanks!

rthadur commented:
approved by @tensorflow/api-owners internally , trying again to merge this internally

mihaimaruseac reviewed (changes requested):
You need to regenerate the API goldens as you changed the API. That's why Ubuntu CPU fails and that's why this does not land. See https://source.cloud.google.com/results/invocations/6ac56c56-69eb-4a88-9020-62ef5fb43a17/log

```
Running tests under Python 3.6.10: /usr/bin/python3
[ RUN      ] ApiCompatibilityTest.testAPIBackwardsCompatibility
/usr/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
  return f(*args, **kwds)
/usr/lib/python3.6/importlib/_bootstrap.py:219: ImportWarning: can't resolve package from __spec__ or __package__, falling back on __name__ and __path__
  return f(*args, **kwds)
ERROR:tensorflow:TensorFlow API backwards compatibility test
This test ensures all changes to the public API of TensorFlow are intended.

If this test fails, it means a change has been made to the public API. Backwards
incompatible changes are not allowed. You can run the test as follows to update
test goldens and package them with your change.

    $ bazel run tensorflow/tools/api/tests:api_compatibility_test \
    #     -- --update_goldens True

You will need an API approval to make changes to the public TensorFlow API. This
includes additions to the API.

E0512 17:59:28.341844 140552551331584 api_compatibility_test.py:293] TensorFlow API backwards compatibility test
This test ensures all changes to the public API of TensorFlow are intended.

If this test fails, it means a change has been made to the public API. Backwards
incompatible changes are not allowed. You can run the test as follows to update
test goldens and package them with your change.

    $ bazel run tensorflow/tools/api/tests:api_compatibility_test \
    #     -- --update_goldens True

You will need an API approval to make changes to the public TensorFlow API. This
includes additions to the API.

ERROR:tensorflow:2 differences found between API and golden.
E0512 17:59:28.342252 140552551331584 api_compatibility_test.py:294] 2 differences found between API and golden.
ERROR:tensorflow:    Change detected in python object: tensorflow.keras.activations.
E0512 17:59:28.342407 140552551331584 api_compatibility_test.py:315]     Change detected in python object: tensorflow.keras.activations.
ERROR:tensorflow:
  path: "tensorflow.keras.activations"
  tf_module {
    member_method {
....
```

gbaned commented:
@sangireddysiva Can you please check @mihaimaruseac's comments and keep us posted ? Thanks!

haifeng-jin commented:
I am closing the PR.
We are trying minimize the redundancy between the core tf API and the keras API.
You can always import it from the tf APIs.

haifeng-jin commented:
@sangireddysiva To minimize the redundancy between the Keras API and the tensorflow ops,
the best solution should be remove the Keras layer but keep the tf ops.

fchollet commented:
@gbaned this PR adds a new activation function in `tf.nn`. Let's ask a TF core team member to review it.

gbaned commented:
> @gbaned this PR adds a new activation function in `tf.nn`. Let's ask a TF core team member to review it.

@fchollet Thank you for the update.

allenlavoie commented on tensorflow/python/ops/nn_impl.py:
```diff
@@ -550,6 +550,48 @@ def grad(dy):
+    # tanh(exp(x)) around for backprop, effectively doubling the tensor's 
+    # memory consumption. We use a control dependency here so that 
+    # tanh(exp(x)) is re-computed during backprop (the control dep prevents 
+    # it being de-duped with the forward pass) and we can free the 
```
Why not `tf.recompute_grad`? That will save a bit more memory executing eagerly and avoids repeating the gradient definitions.

Please also add doctested examples to the tf.nn symbol and a unit test. If you're defining a custom gradient it'd need numeric gradient checking too (probably fine to omit for recompute_grad).

gbaned commented:
@sangireddysiva  Can you please check @allenlavoie's comments and keep us posted ? Thanks!

tensorflowbutler commented:
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

qlzh727 commented:
Removing myself as reviewer since the Keras code has been moved to github.com/keras-team/keras. Please split the keras change into a new PR and sent it to keras-team, if you are still interested in contributing this change. Thanks.

gbaned commented:
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!
